### DISCUSSION TITLE
trees as trees

### DISCUSSION THREAD

**Discussion by justin2004** [saulpw/visidata#2282] [Ideas]
hello all,

i know that visidata's presentation mode is rows and columns but i feel like there could be a more "whole picture" approach to trees. right now when you look at tree like things (directories, xml, json, etc.) you look at one layer at a time and jump in to another layer as needed.

but what if we looked at the whole tree at once?
e.g.

```bash
~/github/visidata$ tree . > /tmp/vd_dir.txt
vd /tmp/vd_dir.txt
```
![image](https://github.com/saulpw/visidata/assets/16325158/91269d52-8009-4a33-b4bf-1fb1aed4d734)

but then each row on its own doesn't give you the context (about what are its parent and its children).
that could be addressed with

```bash
~/github/visidata$ tree -f . > /tmp/vd_dir_f.txt
~/github/visidata$ vd  /tmp/vd_dir_f.txt
```
![image](https://github.com/saulpw/visidata/assets/16325158/788810f6-679d-4dfb-a182-3b3228b095f3)

`vd --dir-recurse .` isn't bad:
![image](https://github.com/saulpw/visidata/assets/16325158/25c35b5f-3776-40d3-9654-a7f94e212eda)

but that kind of display doesn't work for for other tree file types (json, xml, yaml, etc.)

 
i am wondering if anyone has taken this further already?
it would be nice to be able to see trees as trees (instead of just a slice of the tree).

**Comment by midichef** [comment]
@justin2004
I would like to see a tree tool as well! It would be very useful to have tree operations like, "select all siblings", or "select elements whose descendants match a condition".

If you do make a tree tool, I would be happy to try it.

**Comment by justin2004** [comment]
I've made a little progress...
We can now fold and un-fold nodes in the tree. And jump to the parent node.

![image](https://github.com/saulpw/visidata/assets/16325158/6fd80f39-b206-4fb1-8191-300d0f2b17e9)

Let me know if anyone is interested in using this and giving some feedback.

**Reply by saulpw** [reply]
Wow @justin2004, this looks great!  If you post a PR I would love to play with it myself.

**Reply by midichef** [reply]
Folding sounds cool. This looks like a great start!
